Sources/SwiftGRPC/Core/Channel.swift

```diff
@@ -86,83 +85,77 @@ public class Channel {
-    var observer = observers.first(where: { $0.state == sourceState })
-
-    if observer == nil {
-      let newObserver = ConnectivityObserver(state: sourceState, underlyingChannel: underlyingChannel, tryToConnect: tryToConnect)
-      observers.append(newObserver)
-      observer = newObserver
-    }
-
-    observer?.callbacks.append(callback)
-    observer?.polling = true
+  public func subscribe(callback: @escaping (ConnectivityState) -> ()) {
+    let observer = ConnectivityObserver(underlyingChannel: underlyingChannel, callback: callback)
+    observer.polling = true
+    connectivityObservers.append(observer)
   }
 }
 
 private extension Channel {
-  class ConnectivityObserver: Equatable {
-    let state: ConnectivityState
-    let queue: CompletionQueue
-    let underlyingChannel: UnsafeMutableRawPointer
-    let underlyingCompletionQueue: UnsafeMutableRawPointer
-    private(set) var tryToConnect: Bool
-    var callbacks: [(ConnectivityState) -> ()] = []
+  class ConnectivityObserver {
+    private let completionQueue: CompletionQueue
+    private let underlyingChannel: UnsafeMutableRawPointer
+    private let underlyingCompletionQueue: UnsafeMutableRawPointer
+    private let callback: (ConnectivityState) -> ()
     private var lastState: ConnectivityState
-
+    private let queue: OperationQueue
+    
     var polling: Bool = false {
       didSet {
-        if polling == true && oldValue == false {
-          run()
+        queue.addOperation { [weak self] in
+          guard let `self` = self else { return }
+          
+          if self.polling == true && oldValue == false {
+            self.run()
+          } else if self.polling == false && oldValue == true {
+            self.shutdown()
+          }
         }
       }
     }
 
-    init(state: ConnectivityState, underlyingChannel: UnsafeMutableRawPointer, tryToConnect: Bool) {
-      self.state = state
+    init(underlyingChannel: UnsafeMutableRawPointer, callback: @escaping (ConnectivityState) -> ()) {
       self.underlyingChannel = underlyingChannel
-      self.tryToConnect = tryToConnect
       self.underlyingCompletionQueue = cgrpc_completion_queue_create_for_next()
-      self.queue = CompletionQueue(underlyingCompletionQueue: self.underlyingCompletionQueue, name: "Connectivity State")
+      self.completionQueue = CompletionQueue(underlyingCompletionQueue: self.underlyingCompletionQueue, name: "Connectivity State")
+      self.callback = callback
       self.lastState = ConnectivityState.connectivityState(cgrpc_channel_check_connectivity_state(self.underlyingChannel, 0))
+      
+      queue = OperationQueue()
+      queue.maxConcurrentOperationCount = 1
+      queue.qualityOfService = .background
     }
 
     deinit {
-      queue.shutdown()
+      shutdown()
     }
 
     private func run() {
       DispatchQueue.global().async { [weak self] in
```

Comment:
Please use the thread-spawning pattern now used in e.g. https://github.com/grpc/grpc-swift/blob/8941a852aaec9cacf07c6cacc6a6dbbe65c42619/Sources/SwiftGRPC/Core/CompletionQueue.swift#L118 here now instead of dispatching to the global queue.

Reply:
Done